### PR TITLE
Add name kwarg to from_zarr (#1)

### PR DESCRIPTION
**Pull request by mpeaton**
* Add name kwarg to from_zarr

* added default from_zarr naming convention per jrbourbeau

* flake8 fix

- [ na] Tests added / passed
- [ x] Passes `flake8 dask`

### CODE REVIEW

**Comment by jrbourbeau** [comment]
For some context, currently the `from_zarr` function doesn't always produce dask arrays with a unique `.name` attribute. This is because `from_zarr` uses the input zarr array repr to generate the output dask array `name`. For example,

```python
import zarr
import dask
import dask.array as da

a = zarr.array([1, 2, 3])
b = zarr.array([4, 5, 6])
a_dsk = da.from_zarr(a)
b_dsk = da.from_zarr(b)
print(a_dsk.name)
print(b_dsk.name)
print(dask.base.tokenize(a_dsk) == dask.base.tokenize(b_dsk))
```

produces the following output on the current `master`:

```
zarr-<zarr.core.Array (3,) int64>
zarr-<zarr.core.Array (3,) int64>
True
```

The goal of this PR is to provide a fix such that `from_zarr` provides unique dask array names.

**Comment by jrbourbeau** [comment]
cc @jakirkham as you may be interested based on your discussion in https://github.com/zarr-developers/zarr/issues/202

**Comment by mpeaton** [comment]
I have run such a test locally.  The capability is innate to the from-array module, so in the original pass-through design a test was not needed.

**Comment by jrbourbeau** [comment]
I agree that `from_zarr` ultimately passes `name` to `from_array`. I was suggesting to test the default case for `from_zarr` where `name` is `None` and a unique value for `name` is generated using `tokenize`. By running locally I can see that a unique `name` is indeed created, but a short test along the lines of

```python
def test_from_zarr_unique_name():
    a = zarr.array([1, 2, 3])
    b = zarr.array([4, 5, 6])

    assert da.from_zarr(a).name != da.from_zarr(b).name
```

would act as a safeguard to ensure that this bug doesn't pop up again in the future. 

In addition, since this PR adds the new `name` keyword to the `from_zarr` function signature, another short test that makes sure `name` is properly passed to `from_array` when specified would also be useful here. For example:

```python
def test_from_zarr_name():
    a = zarr.array([1, 2, 3])
    assert da.from_zarr(a, name='foo').name == 'foo'
```

File: continuous_integration/appveyor/setup_conda_environment.cmd [hunk 1]
```diff
@@ -26,7 +26,8 @@ copy NUL %CONDA_PREFIX%\conda-meta\pinned
 @rem Install optional dependencies for tests
 %CONDA_INSTALL% numpy pandas cloudpickle distributed
 %CONDA_INSTALL% bcolz bokeh h5py ipython lz4 psutil pytables s3fs scipy
-%CONDA_INSTALL% -c conda-forge fastparquet snappy
+%CONDA_INSTALL% -c conda-forge fastparquet snappy zarr
```
Comment (martindurant) [inline continuous_integration/appveyor/setup_conda_environment.cmd]: python-snappy ?

File: continuous_integration/appveyor/setup_conda_environment.cmd [hunk 2]
```diff
@@ -26,7 +26,8 @@ copy NUL %CONDA_PREFIX%\conda-meta\pinned
...
+%CONDA_INSTALL% asciitree
```
Comment (martindurant) [inline continuous_integration/appveyor/setup_conda_environment.cmd]: Why was this needed?

**Comment by martindurant** [comment]
You may want to rebase on master to take those image and arrow xfails out of the diff

File: continuous_integration/appveyor/setup_conda_environment.cmd (hunk 2 above)
Comment (mpeaton) [reply continuous_integration/appveyor/setup_conda_environment.cmd]: As far as I know zarr is need for test_zarr and asciitree appears to be missing according to the appveyor error report following zarr installation.

**Comment by mpeaton** [comment]
Why is appveryor having trouble with asciitree?

ModuleNotFoundError: No module named 'asciitree'
!!!!!!!!!!!!!!!!!!! Interrupted: 1 errors during collection !!!!!!!!!!!!!!!!!!!
==================== 10 skipped, 1 error in 10.20 seconds =====================

File: dask/array/core.py [hunk 3]
```diff
@@ -2170,6 +2170,12 @@ def from_zarr(url, component=None, storage_options=None, chunks=None, **kwargs):
         Passed to ``da.from_array``, allows setting the chunks on
         initialisation, if the chunking scheme in the on-disc dataset is not
         optimal for the calculations to follow.
+    name : str, optional
+        The key name to use for the array. Defaults to a hash of ``x``.
```
Comment (jrbourbeau) [inline dask/array/core.py]: This description can be shortened and references to `x` (from the `from_array` docstring) should be removed. @mpeaton what do you think about something like "The key name to use for the array. If not specified a unique name will generated."?

File: dask/array/tests/test_array_core.py [hunk 4]
```diff
@@ -3437,6 +3437,30 @@ def test_normalize_chunks_nan():
     assert "auto" in str(info.value)
 
 
+def test_from_zarr_unique_name():
```
Comment (jrbourbeau) [inline dask/array/tests/test_array_core.py]: This test now appears three times

**Comment by jrbourbeau** [comment]
The test failures showing up on Travis CI are related to a recent bokeh update and have been fixed in #4680. 

Other than updating the `name` docstring description (ref https://github.com/dask/dask/pull/4663#discussion_r272712755), this PR looks good to me

**Comment by mrocklin** [comment]
Checking in here.  Is there anything left to do on this PR?

**Comment by jrbourbeau** [comment]
There is one lingering (minor) comment about modifying the `name` docstring description (see https://github.com/dask/dask/pull/4663#discussion_r272712755). @mpeaton do you have any comments on the proposed description?

**Comment by jrbourbeau** [comment]
Thanks, I like your suggestion @mpeaton. One very minor comment, since there is no input parameter `x` for `from_zarr`, could we say "a hash of the input zarr array" instead of "a hash of `x`"? 

Also, just as a note, currently the output from `tokenize(z, ...)` when `z` is a zarr array is random string. It corresponds to `uuid.uuid4().hex` returned here

https://github.com/dask/dask/blob/a2ac8945d58109fc1f8ba1377aa2747d37a1be11/dask/base.py#L642

**Comment by mpeaton** [comment]
@jrbourbeau I see.  Would it be concise, yet accurate enough to replace x with url?
Which raises the question:  What happens in the case where isinstance(url, str)?

**Comment by jrbourbeau** [comment]
>  Would it be concise, yet accurate enough to replace x with url?

Thinking about this again, perhaps the most accurate, yet concise, description would be a general statement about hashing the input (since we're `tokenize`-ing all the `from_zarr` inputs like `component` and `chunks`). For example, the `name` description in the `DataFrame.from_pandas` function is:

https://github.com/dask/dask/blob/91ce8c87ffc992b5b96e1ef2f1971ebefe6d4d73/dask/dataframe/io/io.py#L134-L135

A similar statement would work here too. "An optional key name for the array.  Defaults to hashing the input." Apologies for going back and forth on this point. 


>  Does the mapper immediately evaluate the mapper or is it a lazy reference?

Data should be loaded lazily

File: dask/array/core.py [hunk 5]
```diff
@@ -2170,6 +2170,8 @@ def from_zarr(url, component=None, storage_options=None, chunks=None, **kwargs):
         Passed to ``da.from_array``, allows setting the chunks on
         initialisation, if the chunking scheme in the on-disc dataset is not
         optimal for the calculations to follow.
+    name : str, optional
+         An optional keyname for the dataframe.  Defaults to hashing the input
```
Comment (jrbourbeau) [inline dask/array/core.py]:
```suggestion
         An optional keyname for the array.  Defaults to hashing the input
```

File: continuous_integration/appveyor/setup_conda_environment.cmd [hunk 6]
```diff
@@ -26,7 +26,7 @@ copy NUL %CONDA_PREFIX%\conda-meta\pinned
 @rem Install optional dependencies for tests
 %CONDA_INSTALL% numpy pandas cloudpickle distributed
 %CONDA_INSTALL% bcolz bokeh h5py ipython lz4 psutil pytables s3fs scipy
-%CONDA_INSTALL% -c conda-forge fastparquet snappy
+%CONDA_INSTALL% -c conda-forge fastparquet snappy zarr
```
Comment (jrbourbeau) [inline continuous_integration/appveyor/setup_conda_environment.cmd]: Based on the appveyor logs it looks like we're able to successfully install `zarr` from conda-forge, however the `zarr`-realted tests are being skipped due to not being able to import `zarr` (ref https://ci.appveyor.com/project/daskdev/dask/builds/23862225#L1141). Not sure exactly why this is happening, @jakirkham do you have any thoughts on what could be causing this?

Comment (mpeaton) [reply continuous_integration/appveyor/setup_conda_environment.cmd]: I noted from the error logs that some of the dependencies of zarr did not appear to install properly.

Comment (jrbourbeau) [reply continuous_integration/appveyor/setup_conda_environment.cmd]: Opened up #4706 to track and handle such issues

File: continuous_integration/appveyor/setup_conda_environment.cmd (hunk 1 above)
Comment (jrbourbeau) [reply continuous_integration/appveyor/setup_conda_environment.cmd]: This seems to be the last remaining comment on this PR. @mpeaton would you mind updating this to use `python-snappy` instead of `snappy`? (`python-snappy` will install the `snappy` library along with Python bindings)

File: continuous_integration/appveyor/setup_conda_environment.cmd (hunk 6 above)
Comment (jakirkham) [reply continuous_integration/appveyor/setup_conda_environment.cmd]: Nothing comes to mind. If something does, I'll add it to that issue.

Comment (jrbourbeau) [reply continuous_integration/appveyor/setup_conda_environment.cmd]: Great, thank you for taking a look @jakirkham!

**Comment by jrbourbeau** [comment]
Thanks @mpeaton! I noticed this was your first code contribution to this repository, welcome!
